### PR TITLE
Fix remove overlay

### PR DESCRIPTION
**Pull request by gfgtdf**
fixes LOYAL overlay disappearing from units after levelup.

### CODE REVIEW

**Comment by newfrenchy83** [comment]
i had think to use this method of insert this in traits but i tried other method before.

File: data/core/macros/deprecated-utils.cfg [hunk 1]
```diff
@@ -164,7 +164,7 @@ _"No gold carried over to the next scenario."#enddef
             {TRAIT_UNDEAD}
             {TRAIT_LOYAL}
         [/modifications]
-        overlays="misc/loyal-icon.png"
+        {IS_LOYAL}
```
Comment (stevecotton) [inline data/core/macros/deprecated-utils.cfg]: Does IS_LOYAL do anything useful any more?

If the unit already has TRAIT_LOYAL then that will add the overlay, and if it doesn't then IS_LOYAL will be a bug-in-waiting with the overlay disappearing when the unit levels up.

File: data/campaigns/The_South_Guard/scenarios/03_A_Desparate_Errand.cfg [hunk 2]
```diff
@@ -550,7 +550,7 @@
             name=EthilielV.modifications
             mode=insert
             [value]
-                {TRAIT_LOYAL}
+                {TRAIT_LOYAL OVERLAY=""}
```
Comment (stevecotton) [inline data/campaigns/The_South_Guard/scenarios/03_A_Desparate_Errand.cfg]: I can't justify it, but my gut feeling is that TRAIT_LOYAL_NO_OVERLAY feels better than a macro with an argument.

**Comment by stevecotton** [comment]
I think it would make sense to deprecate support for `[unit]overlay=`, and then verify with the Schema validator that it's not used in mainline any more.

File: data/core/macros/deprecated-utils.cfg (hunk 1 above)
Comment (gfgtdf) [reply data/core/macros/deprecated-utils.cfg]: in the current state of the pr it still adds the ´overlays="misc/loyal-icon.png"´ line, this is wip though, so im not 100% sure how i want ti to be in the end,

**Comment by gfgtdf** [comment]
> I think it would make sense to deprecate support for [unit]overlay=, and then verify with the Schema validator that it's not used in mainline any more.

Hmm while we could in theory do this, this could mean i behaves quite differently from  all other attributes which can be set from both effects and via [unit] attribute (alkthough i'd say you usually shodul prefer the [effect]  way).

File: data/campaigns/The_South_Guard/scenarios/03_A_Desparate_Errand.cfg (hunk 2 above)
Comment (gfgtdf) [reply data/campaigns/The_South_Guard/scenarios/03_A_Desparate_Errand.cfg]: we could, i don't have a strong opinion on it.

**Comment by gfgtdf** [comment]
updated the is with changes from wesnoth:master.

**Comment by CelticMinstrel** [comment]
Confused, isn't this obsoleted by #4211? Am I missing something here?

**Comment by gfgtdf** [comment]
Hmm have to think about it in particular whether I still want to do the change to the loyal trait.

**Comment by CelticMinstrel** [comment]
`[remove_trait]` is passed `trait_id` but is checking for `object_id`.

**Comment by gfgtdf** [comment]
i want to add a duration= key to [unit_overlay] that basiciall sets the duration of the object. any opinion whethe it should be called duration= or object_duration=

**Comment by stevecotton** [comment]
I'd prefer `duration=`, because:
* without the knowledge that `[unit_overlay]` is implemented with `[object]`, there's no reason to expect `object_duration=`
* with that knowledge, it makes sense to be the same key as is used in `[object]`

File: data/campaigns/Sceptre_of_Fire/scenarios/4_Gathering_Materials.cfg [hunk 3]
```diff
@@ -964,10 +969,10 @@
             find_vacant=no
         [/unstore_unit]
 
-        [remove_unit_overlay]
+        [remove_object]
```
Comment (stevecotton) [inline data/campaigns/Sceptre_of_Fire/scenarios/4_Gathering_Materials.cfg]: If the overhead of remove_unit_overlay is so high that it needs to be changed to remove_object, then that feels more like a bug in remove_unit_overlay than something to change in campaigns' files.

Comment (gfgtdf) [reply data/campaigns/Sceptre_of_Fire/scenarios/4_Gathering_Materials.cfg]: i don't think it's a problem real problem it just feels better knowing that there are no useless objects laying around in the unit.

Comment (stevecotton) [reply data/campaigns/Sceptre_of_Fire/scenarios/4_Gathering_Materials.cfg]: If it's worth doing, I'd prefer it to be done in a way that benefits UMC too. Maybe in the unit.cpp's expire_modifications.

I won't have time until Sunday, but would be happy to do the C++ implementation on Sunday.

Comment (gfgtdf) [reply data/campaigns/Sceptre_of_Fire/scenarios/4_Gathering_Materials.cfg]: > If it's worth doing, I'd prefer it to be done in a way that benefits UMC too. Maybe in the unit.cpp's expire_modifications.

i don't think its worth doing this in the c++ code, i mean i only possibility i see is to search for objects that cancel each other out but, then that'd break umc code that relied on such objects and wanted to remove the second object manually via [remove_object].. . I just don't think there is a good way. These wml changes i did were cheap and without disadvantages that's why i did them, changing it in the c++ code has the chance to break other code and for a minimal gain of reducing savefile size of a few bytes.

File: data/core/macros/image-utils.cfg [hunk 4]
```diff
@@ -173,6 +175,71 @@
...
+    [/remove_trait]
+    [modify_unit]
+        [filter]
+            id={ID_STRING}
+        [/filter]
+        canrecruit=yes
+    [/modify_unit]
+#enddef
+
+#define MAKE_LOYAL_HERO IS_STRING
```
Comment (stevecotton) [inline data/core/macros/image-utils.cfg]: `ID`

File: data/campaigns/Sceptre_of_Fire/scenarios/4_Gathering_Materials.cfg (hunk 3 above)
Comment (CelticMinstrel) [reply data/campaigns/Sceptre_of_Fire/scenarios/4_Gathering_Materials.cfg]: I think it should:

1. Make `[remove_unit_overlay]` be no worse than `[remove_object]`.
2. Deprecate `[remove_unit_overlay]` (possibly also `[unit_overlay]`).
3. Replace all mainline uses of `[remove_unit_overlay]` with `[remove_object]`.

File: data/lua/wml/modify_unit.lua [hunk 5]
```diff
@@ -135,6 +136,11 @@ local function simple_modify_unit(cfg)
 				end
 				u:add_modification(tagname, tagcontent);
 			end
+			if tagname == "status" then
```
Comment (CelticMinstrel) [inline data/lua/wml/modify_unit.lua]: Why is this needed? Isn't it handled by the general case?

File: data/campaigns/Sceptre_of_Fire/scenarios/4_Gathering_Materials.cfg (hunk 3 above)
Comment (stevecotton) [reply data/campaigns/Sceptre_of_Fire/scenarios/4_Gathering_Materials.cfg]: If we don't deprecate the tag, then I think it should still be used in mainline, otherwise it won't be tested by playing the mainline campaigns.

File: data/lua/wml/modify_unit.lua (hunk 5 above)
Comment (gfgtdf) [reply data/lua/wml/modify_unit.lua]: well this is,  like the whole simple_modify_unit function kinda an optimisation over the 'general' case which does store+unstore_unit.

**Comment by gfgtdf** [comment]
While doing this i was wondering should heros units (=units with the hero overlay that cost no upkeep but cannot recruit and where you loose when they die) have the 'loyal' trait listed in the traits list or it that not needed?

**Comment by gfgtdf** [comment]
I was again looking at the code and it seems like thre are indeed some cases where {IS_HERO} is given without the loyal trait, but i have no idea why the loyal_trait is omitted there, see for example

https://github.com/wesnoth/wesnoth/blob/a1f855a0a858b422c09e1bd34713cad822b1d38e/data/campaigns/Sceptre_of_Fire/scenarios/1_A_Bargain_is_Struck.cfg#L332

https://github.com/wesnoth/wesnoth/blob/9c1d69254de1111ae051770e623e837bb080ee66/data/campaigns/Delfadors_Memoirs/utils/characters.cfg#L90

**Comment by jostephd** [comment]
@gfgtdf Should this be added to changelog?

**Comment by gfgtdf** [comment]
> @gfgtdf Should this be added to changelog?

Yes, definiteley this is a change of api that might in some cases require addons to update accordingly, it should have a line in the compatability breaking changes section of the 1.15.2 accnouncement  (the 'added overlay effect to the loyal trait' part)


I am in fact unsure whether we should change it again: instead of having 3 different traits with the id loyal, make it only loyal trait that changes the overlay depending on the units [variables].

The reason for that is that i'm having second thoujghts about the 'differnt traits the the same id' concep, even though they only differn by a visual effect and even though it does not casue problerm right now. Mostly for having things like https://github.com/wesnoth/wesnoth/issues/3865 which basicialyl proposes to have a dictionaly indicied by the traits id in lua to give access to the trait definition.

**Comment by jostephd** [comment]
> Yes, definiteley this is a change of api that might in some cases require addons to update accordingly, it should have a line in the compatability breaking changes section of the 1.15.2 accnouncement (the 'added overlay effect to the loyal trait' part)

Then please add it to changelog, it's your PR :)

Same question for 74c2398f67cb904b513733a75825c8dd5867b451.

> I am in fact unsure whether we should change it again: instead of having 3 different traits with the id loyal, make it only loyal trait that changes the overlay depending on the units [variables].

I don't know, sorry. So long as whatever we do is either backward compatible, or failing that, documented as a breaking change, I suppose...

**Comment by jostephd** [comment]
> Then please add it to changelog, it's your PR :)
> 
> Same question for [74c2398](https://github.com/wesnoth/wesnoth/commit/74c2398f67cb904b513733a75825c8dd5867b451).

@gfgtdf Ping?
